sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteSchemaFactory.java

```diff
@@ -17,28 +17,164 @@
  */
 package org.apache.beam.sdk.extensions.sql.impl;
 
+import com.google.common.collect.ImmutableMap;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.Map;
+import java.util.Properties;
 import java.util.ServiceLoader;
+import java.util.Set;
+import org.apache.beam.sdk.extensions.sql.meta.provider.ReadOnlyTableProvider;
 import org.apache.beam.sdk.extensions.sql.meta.provider.TableProvider;
 import org.apache.beam.sdk.extensions.sql.meta.store.InMemoryMetaStore;
 import org.apache.beam.sdk.extensions.sql.meta.store.MetaStore;
+import org.apache.calcite.jdbc.CalciteConnection;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.rel.type.RelProtoDataType;
+import org.apache.calcite.schema.Function;
 import org.apache.calcite.schema.Schema;
 import org.apache.calcite.schema.SchemaFactory;
 import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.SchemaVersion;
+import org.apache.calcite.schema.Table;
 
-/** Factory that creates a {@link BeamCalciteSchema}. */
-public class BeamCalciteSchemaFactory implements SchemaFactory {
-  public static final BeamCalciteSchemaFactory INSTANCE = new BeamCalciteSchemaFactory();
+/**
+ * Factory classes that Calcite uses to create initial schema for JDBC connection.
+ *
+ * <p>Calcite creates a factory using default constructor and isn't flexible enough to hook into the
+ * process to pass in additional parameters (e.g. connection), so we have to add the indirection
+ * layer.
+ *
+ * <p>{@link LoadAllProviders} is the default factory that greedily finds and loads all available
+ * table providers. This is used in a normal JDBC path, e.g. when CLI connects to {@link
+ * JdbcDriver}.
+ *
+ * <p>{@link EmptySchema} is used by {@link JdbcDriver#connect(TableProvider)} to avoid loading all
+ * available table providers.
+ */
+public class BeamCalciteSchemaFactory {
+
+  /**
+   * Called by {@link JdbcConnection} when initializing to convert the initial empty schema to
+   * actual {@link BeamCalciteSchema}.
+   */
+  static TableProvider fromInitialEmptySchema(JdbcConnection jdbcConnection) {
+    InitialEmptySchema initialEmptySchema = jdbcConnection.getCurrentBeamSchema();
+    return initialEmptySchema.toTableProvider(jdbcConnection);
+  }
+
+  /**
+   * Loads all table providers using service loader. This is the default configured in {@link
+   * JdbcDriver#connect(String, Properties)}.
+   */
+  public static class LoadAllProviders extends InitialEmptySchema implements SchemaFactory {
```

Comment:
nit: I would prefer to name a class as a noun. So `ProvidersLoader` or `AllProvidersLoader` might be better.

Reply:
I'll think about naming. The intent here is to emphasize the action in `JdbcDriver`. At that point it's clear that we're using `BeamCalciteSchemaFactory` to produce schema, but these nested classes describe the specific behavior, i.e. `BeamCalciteSchemaFactory.LoadAllProviders`, so that it reads like a verb. Ideally I wanted a factory method there instead but I don't believe it's supported by Calcite

Reply:
renamed to `AllProviders` and `Empty`